src/NServiceBus.RabbitMQ/Receiving/MessagePump.cs

```diff
@@ -165,6 +167,30 @@ void Connection_ConnectionShutdown(object sender, ShutdownEventArgs e)
 
             try
             {
+                // The current thread will be the event raising thread if either:
+                //
+                // a) the semaphore was entered synchronously (did not have to wait).
+                // b) the event was raised on a thread pool thread,
+                //    and the semaphore was entered asynchronously (had to wait),
+                //    and the continuation happened to be scheduled back onto the same thread.
+                if (Thread.CurrentThread.ManagedThreadId == eventRaisingThreadId)
+                {
+                    // In RabbitMQ.Client 4.1.0, the event is raised by reusing a single explicitly created thread
+                    // so we are in scenario (a) described above.
+                    // We must yield to allow the thread to raise more events while we handle this one,
+                    // otherwise we will never process messages concurrently.
+                    //
+                    // If a future version of RabbitMQ.Client changes it's threading model then either:
+                    //
+                    // 1) we are in scenario (a) but we *may not* need to yield.
+                    //    E.g. the client may raise the event on a new explicitly created thread each time.
+                    // 2) we cannot tell whether we are in scenario (a) or scenario (b).
+                    //    E.g. the client may raise the event on thread pool thread.
+                    // 
+                    // In both these cases we cannot tell whether we need to yield or not, so we must yield.
```

Comment:
I believe [both forms are correct](http://ell.stackexchange.com/questions/110665/both-of-these-or-both-these) but I don't have a strong preference. I have a weak preference for omitting "of".